dynamiqs/mesolve/rouchon.py

```diff
@@ -53,26 +44,55 @@ def forward(self, t: float, dt: float, rho: Tensor) -> Tensor:
         """
         # rho: (b_H, b_rho, n, n) -> (b_H, b_rho, n, n)
 
+        # compute cached operators
+        H = self.H(t)  # (b_H, 1, n, n)
+        H_nh = self.H_nh(H)  # (b_H, 1, n, n)
+        M0 = self.M0(H_nh)  # (b_H, 1, n, n)
```

Comment:
Same comment as before about grouping lines.

Reply:
Also for someone who wants to read along and understand how the caching works nothing is hidden: `H` is a new tensor only if `self.H(t)` returns something new, and the subsequent tensors are only re-computed if `H` is a new tensor.

Reply:
Yes let's keep it like you wrote already. I don't think there should be any performance/memory gain by one-lining this type of code in Python. There could be however if it was something like
```
H = A
H += B
H *= c
```
because operations would be fused (I think, not 100% sure).